Prevent broken `Content-Type` from raising exceptions

kirs opened the pull request:
My app gets requests from bots like this one:

```
{
  "REQUEST_METHOD" => "POST",
  "REQUEST_PATH" => "/index.php",
  "REQUEST_URI" => "/index.php?option=com_jce&task=plugin&plugin=imgmanager&file=imgmanager&method=form&cid=20&6bc427c8a7981f4fe1f5ac65c1246b5f=cf6dd3cf1923c950586d0dd595c8e20b",
  "SERVER_PROTOCOL" => "HTTP/1.0",
  "HTTP_VERSION" => "HTTP/1.0",
  "HTTP_HOST" => "example.com",
  "HTTP_X_SENDFILE_TYPE" => "X-Accel-Redirect",
  "HTTP_CONNECTION" => "close",
  "CONTENT_LENGTH" => "2286",
  "HTTP_X_FORWARDED_PROTO" => "http",
  "HTTP_USER_AGENT" => "BOT/0.1 (BOT for JCE)",
  "CONTENT_TYPE" => "multipart/form-data; boundary=---------------------------41184676334",
  "SERVER_NAME" => "example.com",
  "SERVER_PORT" => "80",
  "SERVER_SOFTWARE" => "Unicorn 4.8.3",
  "ORIGINAL_FULLPATH" => "/index.php?option=com_jce&task=plugin&plugin=imgmanager&file=imgmanager&method=form&cid=20&6bc427c8a7981f4fe1f5ac65c1246b5f=cf6dd3cf1923c950586d0dd595c8e20b"
}
```

This requests contents broken `CONTENT_TYPE` value, which leads to uncaught `EOFError` in production thrown by Rack parser.
I guess we should avoid such uncaught exceptions.

Similar with https://github.com/rails/rails/commit/bc254cc23558f016c3697ecd5d39e58f46908018

kirs committed:
Prevent broken CONTENT_TYPE with boundary from returning 500
Mostly to avoid 500s on bot requests
EOFError should not be raise BadRequest on GET request because of IE issue:
https://github.com/rails/rails/commit/bc254cc23558f016c3697ecd5d39e58f46908018

rafaelfranca commented:
Not sure. If the request is broken why should we return a ok response?

matthewd commented:
Yeah.. I could see us translating it into a more informative BadRequest, but we can't just pretend everything was fine.

egilburg commented:
Should this be fixed in Rack itself?

kirs commented:
Updated.
As soon as Rails [already has a fix for IE](https://github.com/rails/rails/commit/bc254cc23558f016c3697ecd5d39e58f46908018), I think it's better to solve the problem in Rails.

In the updated version of PR I simplified the previous fix for IE and updated the syntax for `get` request in tests.

kirs commented:
/cc @matthewd @rafaelfranca

rafaelfranca commented on actionpack/lib/action_dispatch/http/request.rb:
```diff
@@ -318,7 +318,11 @@ def GET
-    rescue Rack::Utils::ParameterTypeError, Rack::Utils::InvalidParameterError => e
+    rescue Rack::Utils::ParameterTypeError, Rack::Utils::InvalidParameterError, EOFError => e
+      if e.is_a?(EOFError) && get?
+        return {}
```
Why not Bad Request too?

kirs replied:
Because of the workaround for IE here https://github.com/rails/rails/commit/bc254cc23558f016c3697ecd5d39e58f46908018

kirs commented:
Any thoughts? @rafaelfranca

kirs commented:
ping @rafaelfranca
